src/os/bluestore/BlueStore.cc

```diff
@@ -5363,706 +5225,466 @@ void BlueStore::_pad_zeros_tail(
-        dout(10) << __func__ << "  final head cow op extent "
-                 << cow_head_op->extent << dendl;
-      }
-      if (e.length == length && cow_tail_op) {
-        // we set cow_tail_extent to indicate that all or part of this
-        // new extent will be copied from the previous allocation.
-	*cow_tail_extent = e.offset;
-        // extent.offset is the logical object offset
-        assert(cow_tail_op->extent.offset >= offset);
-        assert(cow_tail_op->extent.end() <= offset + length);
-        cow_tail_op->extent.offset += e.offset - offset;
-        dout(10) << __func__ << "  final tail cow op extent "
-                 << cow_tail_op->extent << dendl;
-      }
+      need -= l;
+      e.length = l;
       txc->allocated.insert(e.offset, e.length);
-      b->extents.push_back(e);
-      offset += e.length;
+      (*bp)->extents.push_back(e);
       length -= e.length;
       hint = e.end();
     }
-    bluestore_lextent_t& le =
-      o->onode.extent_map[*alloc_offset] = bluestore_lextent_t(
-	blob,
-	0,
-	*alloc_length,
-	0);
-    dout(10) << __func__ << "  alloc 0x" << std::hex << offset << std::dec
-	     << ": " << le << dendl;
-    b->length = *alloc_length;
-    dout(10) << __func__ << "   blob " << blob << " " << *b << dendl;
-  }
+    dout(20) << __func__ << " blob " << **bp << dendl;
 
+    // checksum
+    if (csum_type) {
+      b->init_csum(csum_type, 12, l->length()); // FIXME adjust b size
+      checksummer->calculate(b->csum_type, b->get_csum_block_size(),
+			     0, l->length(), *l,
+			     &b->csum_data);
+    }
+
+    // queue io
+    bufferlist::iterator p = l->begin();
+    for (auto &v : (*bp)->extents) {
+      bufferlist t;
+      p.copy(v.length, t);
+      bdev->aio_write(v.offset, t, &txc->ioc, wctx->buffered);
+    }
+  }
+  if (need > 0) {
+    alloc->unreserve(need);
+  }
   return 0;
 }
 
-bool BlueStore::_can_overlay_write(OnodeRef o, uint64_t length)
-{
-  return
-    (int)o->onode.overlay_map.size() < g_conf->bluestore_overlay_max &&
-    (int)length <= g_conf->bluestore_overlay_max_length;
+void BlueStore::_wctx_finish(
+  TransContext *txc,
+  CollectionRef& c,
+  OnodeRef o,
+  WriteContext *wctx)
+{
+  dout(10) << __func__ << " lex_old " << wctx->lex_old << dendl;
+  for (auto &l : wctx->lex_old) {
+    bluestore_blob_t *b = c->get_blob_ptr(o, l.blob);
+    // fixme: handle partial deallocation
+    b->ref_map.put(l.offset, l.length, NULL);
+    if (b->ref_map.empty()) {
+      dout(20) << __func__ << " rm blob " << *b << dendl;
+      for (auto p : b->extents) {
+	txc->released.insert(p.offset, p.length);
+      }
+      if (l.blob >= 0) {
+	o->onode.blob_map.erase(l.blob);
+      } else {
+	o->bnode->blob_map.erase(-l.blob);
+      }
+    } else {
+      dout(20) << __func__ << " keep blob " << *b << dendl;
+    }
+    if (l.blob >= 0) {
+      txc->write_onode(o);
+    } else {
+      txc->write_bnode(o->bnode);
+    }
+  }
+
+  o->onode.compress_extent_map();
```

Comment:
Suggest to call txc->write_onode(o) if result > 0 here for the sake of code view consistency